### PR TITLE
Fix issue related to duplicate identical bind mounts for `docker run`

### LINKED ISSUE #27969: duplicate identical bind mounts create volumes with colon in the filename (-v /tmp:/tmp -v /tmp:/tmp)
**Description**

When doing `docker run` with bind mounts, specifying the same bind mount twice creates second volume with a colon in the path instead of deduplicating (either at the Docker level or mount-on-mount at the kernel level).

**Steps to reproduce the issue:**
```console
$ docker run -it --rm -v /tmp:/tmp -v /tmp:/tmp bash
bash-4.4# ls -l /
...
drwxrwxrwt    2 root     root            40 Nov  1 22:11 tmp
drwxr-xr-x    3 root     root          4096 Nov  1 22:12 tmp:
...
bash-4.4# mount | grep /tmp
none on /tmp type tmpfs (rw,relatime)
/dev/sdc3 on /tmp:/tmp type ext4 (rw,noatime,nodiratime,data=ordered)
```
(`/tmp` on my host is a `tmpfs`, hence the "real" bind-mounted `/tmp` showing up as `tmpfs` in the container)

**Describe the results you received:**

```json
        "Mounts": [
            {
                "Type": "bind",
                "Source": "/tmp",
                "Destination": "/tmp",
                "Mode": "",
                "RW": true,
                "Propagation": ""
            },
            {
                "Type": "volume",
                "Name": "56dd1395abd0362437119e589c3df80d490a91072abca007c7c3c3974ef1c259",
                "Source": "/var/lib/docker/volumes/56dd1395abd0362437119e589c3df80d490a91072abca007c7c3c3974ef1c259/_data",
                "Destination": "/tmp:/tmp",
                "Driver": "local",
                "Mode": "",
                "RW": true,
                "Propagation": ""
            }
        ],
```

**Describe the results you expected:**

```
        "Mounts": [
            {
                "Type": "bind",
                "Source": "/tmp",
                "Destination": "/tmp",
                "Mode": "",
                "RW": true,
                "Propagation": ""
            }
        ],
```

or:

```
        "Mounts": [
            {
                "Type": "bind",
                "Source": "/tmp",
                "Destination": "/tmp",
                "Mode": "",
                "RW": true,
                "Propagation": ""
            },
            {
                "Type": "bind",
                "Source": "/tmp",
                "Destination": "/tmp",
                "Mode": "",
                "RW": true,
                "Propagation": ""
            }
        ],
```

**Output of `docker version`:**

```console
$ docker version
Client:
 Version:      1.13.0-dev
 API version:  1.25
 Go version:   go1.7.3
 Git commit:   bf16fa4
 Built:        Tue Nov  1 18:41:05 2016
 OS/Arch:      linux/amd64

Server:
 Version:      1.13.0-dev
 API version:  1.25
 Go version:   go1.7.3
 Git commit:   bf16fa4
 Built:        Tue Nov  1 18:41:05 2016
 OS/Arch:      linux/amd64
 Experimental: false
```

(Discovered on 1.12.1, but tested on a master build to ensure it wasn't already fixed :+1:)

**Output of `docker info`:**

```console
$ docker info
Containers: 1
 Running: 1
 Paused: 0
 Stopped: 0
Images: 1
Server Version: 1.13.0-dev
Storage Driver: overlay2
 Backing Filesystem: extfs
Logging Driver: json-file
Cgroup Driver: cgroupfs
Plugins: 
 Volume: local
 Network: bridge host macvlan null overlay
Swarm: inactive
Runtimes: runc
Default Runtime: runc
Security Options: seccomp
Kernel Version: 4.4.23-gentoo
Operating System: Alpine Linux v3.4 (containerized)
OSType: linux
Architecture: x86_64
CPUs: 8
Total Memory: 31.4 GiB
Name: 37229d7b91bd
ID: LONU:IWDA:HKNR:EFWO:TTFF:TPSJ:7HIU:6B2T:7O7J:E4JE:SSTQ:2EHW
Docker Root Dir: /var/lib/docker
Debug Mode (client): false
Debug Mode (server): false
Registry: https://index.docker.io/v1/
WARNING: bridge-nf-call-iptables is disabled
WARNING: bridge-nf-call-ip6tables is disabled
Experimental: false
Insecure Registries:
 127.0.0.0/8
Live Restore Enabled: false
```

**Additional environment details (AWS, VirtualBox, physical, etc.):**

As noted, happens on my host, and in Docker-in-Docker (not really unexpected, but noting for completeness), and both in 1.12 and master.

### PR DESCRIPTION
**Pull request by yongtang**
**- What I did**

This fix tries to address the issue raised in #27969 where duplicate identical bind mounts for `docker run` caused additional volumes to be created.

**- How I did it**

The reason was that in `runconfig`, if duplicate identical bind mounts have been specified, the `copts.volumes.Delete(bind)` will not truly delete the second entry from the slice. (Only the first entry is deleted).

This fix fixes the issue.

**- How to verify it**

An integration test has been added to cover the changes

**- Description for the changelog**

**- A picture of a cute animal (not mandatory but encouraged)**

cute-panda-s-3-pandas-21084038-400-265

This fix fixes #27969.

Signed-off-by: Yong Tang <yong.tang.github@outlook.com>

### CODE REVIEW

File: integration-cli/docker_cli_run_test.go [hunk 1]
```diff
@@ -4567,3 +4567,14 @@ func (s *DockerSuite) TestRunServicingContainer(c *check.C) {
 	c.Assert(out2, checker.Contains, `Windows Container (Servicing)`, check.Commentf("Didn't find 'Windows Container (Servicing): %s", out2))
 	c.Assert(out2, checker.Contains, containerID+"_servicing", check.Commentf("Didn't find '%s_servicing': %s", containerID+"_servicing", out2))
 }
+
+func (s *DockerSuite) TestRunDuplicateMount(c *check.C) {
+	testRequires(c, DaemonIsLinux)
+
+	name := "test"
+	out, _ := dockerCmd(c, "run", "--name", name, "-v", "/tmp:/tmp", "-v", "/tmp:/tmp", "busybox", "ls", "/")
+	c.Assert(out, checker.Not(checker.Contains), "tmp:")
```
Comment (AkihiroSuda) [inline integration-cli/docker_cli_run_test.go]: Can you add an assertion that checks whether `/tmp` is actually bind-mounted?

Maybe you should use another directory with some dummy data inside for the purpose.

**Review by AkihiroSuda — CHANGES_REQUESTED** [review]

**Comment by yongtang** [comment]
Thanks @AkihiroSuda. The PR has been updated.

**Review by AkihiroSuda — APPROVED** [review]
LGTM

**Comment by coolljt0725** [comment]
LGTM

**Comment by AkihiroSuda** [comment]
Merged, thank you

**Comment by tianon** [comment]
That was insanely fast -- thanks folks :heart: :heart:

**Pull request merged**